benchmark/dataset_creation/codebase_chunker.py

```diff
@@ -0,0 +1,147 @@
+            raise FileNotFoundError(f"Codebase directory not found: {self.codebase_path}")
+        
+        python_files = [os.path.join(root, file) 
+                        for root, _, files in os.walk(self.codebase_path) 
+                        for file in files if file.endswith('.py')]
+        
+        for file_path in tqdm(python_files, desc="Chunking codebase", unit="file"):
+            print(f"Chunking: {file_path}")  # Print each file path
+            try:
+                with open(file_path, 'r') as f:
+                    content = f.read()
+                ast_tree = parse_ast(content)
+                file_chunks = self._generate_chunks(ast_tree, file_path)
+                chunks.extend(file_chunks)
+            except FileNotFoundError:
+                print(f"Warning: File not found: {file_path}")
+            except IOError as e:
+                print(f"Error reading file {file_path}: {e}")
+            except Exception as e:
+                print(f"Unexpected error processing file {file_path}: {e}")
+        return chunks
+
+    def _generate_chunks(self, tree: ast.AST, file_path: str) -> List[Dict[str, Any]]:
+        chunks = []
+        
+        # Function-level chunks
+        function_chunks = self._get_function_chunks(tree, file_path)
+        chunks.extend(function_chunks)
+        
+        # File-level chunk
+        file_chunk = self._get_file_chunk(tree, file_path)
+        chunks.append(file_chunk)
+
+        return chunks
+
+    def _get_function_chunks(self, tree: ast.AST, file_path: str) -> List[Dict[str, Any]]:
+        function_chunks = []
+        for node in ast.walk(tree):
+            if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef, ast.ClassDef)):
+                chunk = self._create_chunk(node, file_path)
+                if chunk:
+                    function_chunks.append(chunk)
+        return function_chunks
+
+    def _get_file_chunk(self, tree: ast.AST, file_path: str) -> Dict[str, Any]:
+        return {
+            'type': 'file',
+            'content': ast.unparse(tree),
+            'metadata': {
+                'file': file_path,
+                'name': os.path.basename(file_path)
+            }
+        }
+
+    def _create_chunk(self, node: ast.AST, file_path: str) -> Dict[str, Any]:
+        content = ast.unparse(node)
+        tokens = self.tokenizer.encode(content)
+        
+        if len(tokens) <= self.max_tokens:
+            return {
+                'type': 'function' if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)) else 'class',
+                'content': content,
+                'metadata': {
+                    'file': file_path,
+                    'name': node.name,
+                    'lineno': node.lineno,
+                    'end_lineno': node.end_lineno
+                }
+            }
+        else:
+            # If the chunk is too large, we need to split it
+            return self._split_large_chunk(node, file_path)
```

Comment:
_split_large_chunk returns a List[Dict], should create_chunk also returns a list?

Reply:
huh we could i guess i just parse it differently if it's a dictionary

Reply:
you could parse it differently, but the same function is still returning two different types? Then in line 56:
```
chunk = self._create_chunk(node, file_path)
```

chunk could be either a List[Dict] or Dict. 

```
function_chunks.append(chunk)
```
could be appending a list to a list -> List[Dict, Dict, List[Dict], Dict...]